README
Copyright (c) 1999 Jason Kohles.
Copyright (c) 2001 Ivan Kohler
All rights reserved. This program is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

This is Business::OnlinePayment::AuthorizeNet, an Authorize.Net-specific
backend module for Business::OnlinePayment.  It has been updated and now
implements Authorize.Net API version 3.0.

Jason Kohles is the original author.
Ivan Kohler <ivan-authorizenet@420.am> is the current maintainer.

Business::OnlinePayment is a generic interface for processing payments through
online credit card processors, online check acceptance houses, etc.  (If you
like buzzwords, call it an "multiplatform ecommerce-enabling middleware
solution").


Changes
Revision history for Perl extension Business::OnlinePayment::AuthorizeNet.

3.22  Mon Sep 26 15:06:52 PDT 2011
        - Add missing t/lib/Business/FraudDetect/_Fake.pm to MANIFEST to fix
          tests.  Patch from Rob Brown, thanks!  (closes: CPAN#52444)
        - Incorporate Business::OnlinePayment::AuthorizeNet::AIM::ErrorCodes by
          Thomas Sibley and, using it, provide more descriptive error messages.
          (closes: CPAN#34362)
        - Switch to Business::OnlinePayment::HTTPS instead of using Net::SSLeay
          directly
        - Silence new warnings about lc(undef) introduced in perl 5.12.  Patch
          from Todd Rinaldo, thanks!  (closes: CPAN#56172)

3.21  Tue Nov 24 10:45:21 PST 2009
        - Add repository, contributing and contirbutor information to the docs.
        - Patch from Nate Nuss <ogmoid[...]gmail.com> implementing ("Additional
          Shipping Information (Level 2 Data)" ~pg 24 in the AIM guide):
          tax, freight, duty, tax_exempt, po_number.  Thanks!
          (closes: CPAN#42046)
        - Patch from Michael Peters to fix a bug in email address handling:
          exclude x_Email_Customer flag when it is not specified in content,
          to avoid overriding admin settings.  (closes: CPAN#51501)
        - Patch from Josh Rosenbaum to fix encapsulation problems.  Thanks!
          (closes: CPAN#15210)
        - Clarified documentation wrt transaction key (closes: CPAN#23753)
        - ARB (recurring billing) fixes from John Springer, thanks!
          (closes: CPAN#48625)
        - Add META.yml

3.20  Wed Jun 18 16:46:10 PDT 2008
        - Patch from Erik Hollensbe <erikh@opensourcery.com> implementing
          card-present data (track1/track2) and the duplicate_window parameter,
          and test fixes.  Thanks!
        - Patch from Paul Timmins <paul@timmins.net> adding check_number field.

3.19  Fri Nov 23 12:46:05 PST 2007
        - ironically, forgot the 3.18 changelog, so this is 3.19 anyway :)

3.18  Fri Nov 23 10:55:44 PST 2007
        - Patch From Steve Simitzis for better compatiblity with
          eProcessingNetwork's AuthorizeNet compatability mode.
        - added ARB support, rearranging code in the process (Jeff Finucane)

3.17  Tue Jul 10 21:12:46 PDT 2007
        - Trim the extra 'ip_addr="1.2.3.4"' added by eProcessingNetwork's
          AuthorizeNet compatability mode.
        - add bank account type handling

3.16  Tue Nov 14 02:35:30 PST 2006
	- Update link to API docs, now it is called "Advanced Integration
	  Method (AIM)"
	- Update test account, separate ACH-capable one
	- patch to map ship_company properly to x_Ship_To_Company from Mike
	  Barry <MBarry@cos.com>
	- Documentation patch from William McKee <WMCKEE@cpan.org>
	From Michael G. Schwern <MSCHWERN@cpan.org>:
	- Eliminate inheriting from AutoLoader. We're not using it and it just
          screws up the error messages.
        - account_type mentioned twice in the required fields for checks.
        - Quiet an uninit value warning when customer_org is not set.
	- Fix t/credit_card.t test to use a date 11 months in the future as
	  the expiration date, and to print the error message on failure.
	- Update tests to use Test::More

3.15  Wed Mar 16 01:10:51 PST 2005
	- Ask for ',' delimiter and '"' quote explicitly to prevent problems
	  when a merchant has them configured differently.
	- expiration is not a required field for credits

3.14  Tue Sep 21 01:10:46 PDT 2004
	- Added ship_ name/address fields
	- Fixed required fields for echeck voiding
	From T.J. Mather <tjmather@maxmind.com> (closes: cpan#6761):
	- Added cvv2_response and cavv_response to get the cvv2 and cavv
	  response codes back from authorize.net.
	- Added example of how to capture a transaction to the synopsis.
	- Added example of how to get the avs_code, cvv2_response, and
	  cavv_response from the response

3.13  Sun Aug 10 21:56:34 PDT 2003
	- removed Michael Mavroudis's email address
	- added 'void' action, patch from Yuri V. Mkrtumyan
	  <yuramk@novosoft.ru>
	- allow "Post Authorize" with order_number but no card info, patch
	  from Paul Zimmer <AuthorizeNetpm@pzimmer.box.bepress.com>
	- Don't require "check_type" field
	- s/x_Bank_Account_Name/x_Bank_Acct_Name/ for ACH transactions
	From Daemon Hughes <daemmon@daemmonhughes.com>:
	- add "transaction key" (x_Tran_Key) authentication in addition to
	  password
	- add "recurring_billing" field
	- added "md5" method
	From valerian <valerian2@hotpop.com>:
	- Added "avs_code" method
	- Return "order_number" for unsucessful transactions also

3.12  Thu Nov 21 17:05:19 2002
	- Added cvv2 field, patch from T.J. Mather <tjmather@maxmind.com>
	  (closes: cpan#1805)
	- Added referer field, patch from Mike Barry <mbarry@cos.com>
	- Added currency field, request from
	  Oleksandr Kapitanenko <kapitan@portaone.com>
	- Added documentation about referrer field at Authorize.Net
	- Added nonascii.patch from "T.J. Mather" <tjmather@maxmind.com> to
	  put Text::CSV in binary mode to prevent problems with i18n
	  characters (closes: cpan#1804)
	- ECHECK (ACH) patch from Michael Mavroudis <email removed by
	  request> to add fields for ssn and license #

3.11  Sat May  4 00:43:36 PDT 2002
	- forgot 3.10 changelog in 3.10 :)
	- extremely verbose debugging information for responses without
	  response code
	- PREREQ_PM on Business::OnlinePayment
	- Escape 0x00 (NULL) in responses from Authorize.Net.  wtf?

3.10  Wed Mar 13 2002
	- updated for Authorize.Net API 3.1
	- enable t/credit_card.t test again; testing account seems to work
	- working Post Authorization support
	- s/CSV/CSV_XS/

3.01  Wed Nov 14 13:42:06 2001
	- update README
	- disable t/credit_card.t test; testdrive account no longer valid

3.00  Sat Sep  1 13:29:34 2001
	- new maintainer, updated for Authorize.Net API 3.0

0.01  Sun Jul 25 16:37:11 1999
	- original version; created by h2xs 1.19



t/lib/test_account.pl
sub test_account_or_skip {
    my $suffix = shift;
    my($login, $password) = test_account($suffix);

    unless( defined $login ) {
        plan skip_all => "No test account";
    }

    return($login, $password);
}

sub test_account {
    my $suffix = shift || '';
    $suffix = "_$suffix" if $suffix;
    open TEST_ACCOUNT, "t/test_account$suffix" or return;
    my($login, $password) = <TEST_ACCOUNT>;
    chomp $login;
    chomp $password;

    return($login, $password);
}

sub expiration_date {
    my($month, $year) = (localtime)[4,5];
    $year++;       # So we expire next year.
    $year %= 100;  # y2k?  What's that?

    return sprintf("%02d/%02d", $month, $year);
}

sub tomorrow {
    my($day, $month, $year) = (localtime(time+86400))[3..5];
    return sprintf("%04d-%02d-%02d", $year+1900, ++$month, $day);
}

1;


AuthorizeNet/ARB.pm
package Business::OnlinePayment::AuthorizeNet::ARB;

use strict;
use Carp;
use Business::OnlinePayment::AuthorizeNet;
use Business::OnlinePayment::HTTPS;
use XML::Simple;
use XML::Writer;
use Tie::IxHash;
use vars qw($VERSION $DEBUG @ISA $me);

@ISA = qw(Business::OnlinePayment::AuthorizeNet Business::OnlinePayment::HTTPS);
$VERSION = '0.02';
$DEBUG = 0;
$me='Business::OnlinePayment::AuthorizeNet::ARB';

sub set_defaults {
    my $self = shift;

    $self->server('api.authorize.net') unless $self->server;
    $self->port('443') unless $self->port;
    $self->path('/xml/v1/request.api') unless $self->path;

    $self->build_subs(qw( order_number md5 avs_code cvv2_response
                          cavv_response
                     ));
}

sub map_fields {
    my($self) = @_;

    my %content = $self->content();

    # ACTION MAP
    my %actions = ('recurring authorization'
                      => 'ARBCreateSubscriptionRequest',
                   'modify recurring authorization'
                      => 'ARBUpdateSubscriptionRequest',
                   'cancel recurring authorization'
                      => 'ARBCancelSubscriptionRequest',
                  );
    $content{'action'} = $actions{lc($content{'action'})} || $content{'action'};

    # TYPE MAP
    my %types = ('visa'               => 'CC',
                 'mastercard'         => 'CC',
                 'american express'   => 'CC',
                 'discover'           => 'CC',
                 'check'              => 'ECHECK',
                );
    $content{'type'} = $types{lc($content{'type'})} || $content{'type'};
    $self->transaction_type($content{'type'});

    # ACCOUNT TYPE MAP
    my %account_types = ('personal checking'   => 'checking',
                         'personal savings'    => 'savings',
                         'business checking'   => 'businessChecking',
                         'business savings'    => 'savings',
                        );
    $content{'account_type'} = $account_types{lc($content{'account_type'})}
                               || $content{'account_type'};

    # MASSAGE EXPIRATION 
    $content{'expdate_yyyymm'} = $self->expdate_yyyymm($content{'expiration'});

    # stuff it back into %content
    $self->content(%content);

}

sub revmap_fields {
  my $self = shift;
  tie my(%map), 'Tie::IxHash', @_;
  my %content = $self->content();
  map {
        my $value;
        if ( ref( $map{$_} ) eq 'HASH' ) {
          $value = $map{$_} if ( keys %{ $map{$_} } );
        }elsif( exists( $content{ $map{$_} } ) ) {
          $value = $content{ $map{$_} };
        }

        if (defined($value)) {
          ($_ => $value);
        }else{
          ();
        }
      } (keys %map);
}

sub expdate_yyyymm {
  my $self = shift;
  my $expiration = shift;
  my $expdate_yyyymm;
  if ( defined($expiration) and $expiration =~ /^(\d{1,2})\D+(\d{2})$/ ) {
    my ( $month, $year ) = ( $1, $2 );
   $expdate_yyyymm = sprintf( "20%02d-%02d", $year, $month );
  }
  return defined($expdate_yyyymm) ? $expdate_yyyymm : $expiration;
};

sub _xmlwrite {
  my ($self, $writer, $item, $value) = @_;
  $writer->startTag($item);
  if ( ref( $value ) eq 'HASH' ) {
    foreach ( keys ( %$value ) ) {
      $self->_xmlwrite($writer, $_, $value->{$_});
    }
  }else{
    $writer->characters($value);
  }
  $writer->endTag($item);
}

sub submit {
  my($self) = @_;

  $self->map_fields();

  my @required_fields = qw(action login password);

  if ( $self->{_content}->{action} eq 'ARBCreateSubscriptionRequest' ) {
    push @required_fields,
      qw( type interval start periods amount first_name last_name );

    if ($self->transaction_type() eq "ECHECK") {
      push @required_fields,
        qw( amount routing_code account_number account_type account_name
            check_type
          );
    } elsif ($self->transaction_type() eq 'CC' ) {
      push @required_fields, qw( card_number expiration );
    }
  }elsif ( $self->{_content}->{action} eq 'ARBUpdateSubscriptionRequest' ) {
    push @required_fields, qw( subscription );
  }elsif ( $self->{_content}->{action} eq 'ARBCancelSubscriptionRequest' ) {
    push @required_fields, qw( subscription );
  }else{
    croak "$me can't handle transaction type: ".
      $self->{_content}->{action}. " for ". 
      $self->transaction_type();
  }

  $self->required_fields(@required_fields);

  tie my %merchant, 'Tie::IxHash',
    $self->revmap_fields(
                          name           => 'login',
                          transactionKey => 'password',
                        );

  my ($length,$unit) =
    ($self->{_content}->{interval} or '') =~ /^\s*(\d+)\s+(day|month)s?\s*$/;
  tie my %interval, 'Tie::IxHash', (
                                     ($length ? (length => $length)   : () ),
                                     ($unit   ? (unit   => $unit.'s') : () ),
                                   );

  tie my %schedule, 'Tie::IxHash',
    $self->revmap_fields(
                          interval         => \%interval,
                          startDate        => 'start',
                          totalOccurrences => 'periods',
                          trialOccurrences => 'trialperiods',
                        );

  tie my %account, 'Tie::IxHash', ( 
    ( defined($self->transaction_type())
      && $self->transaction_type() eq 'CC'
    ) ? $self->revmap_fields(
                              cardNumber     => 'card_number',
                              expirationDate => 'expdate_yyyymm',
                            )
      : $self->revmap_fields(
                              accountType    => 'account_type',
                              routingNumber  => 'routing_code',
                              accountNumber  => 'account_number',
                              nameOnAccount  => 'account_name',
                              echeckType     => 'check_type',
                              bankName       => 'bank_name',
                            )
  );

  tie my %payment, 'Tie::IxHash',
    $self->revmap_fields(
                           ( ( defined($self->transaction_type()) && # require?
                               $self->transaction_type() eq 'CC'
                             ) ?  'creditCard'
                               : 'bankAccount'
                           )  => \%account,
                         );

  tie my %order, 'Tie::IxHash',
    $self->revmap_fields(
                          invoiceNumber => 'invoice_number',
                          description   => 'description',
                        );

  tie my %drivers, 'Tie::IxHash',
    $self->revmap_fields(
                          number      => 'license_num',
                          state       => 'license_state',
                          dateOfBirth => 'license_dob',
                        );

  tie my %billto, 'Tie::IxHash',
    $self->revmap_fields(
                          firstName => 'first_name',
                          lastName  => 'last_name',
                          company   => 'company',
                          address   => 'address',
                          city      => 'city',
                          state     => 'state',
                          zip       => 'zip',
                          country   => 'country',
                        );

  tie my %shipto, 'Tie::IxHash',
    $self->revmap_fields(
                          firstName => 'ship_first_name',
                          lastName  => 'ship_last_name',
                          company   => 'ship_company',
                          address   => 'ship_address',
                          city      => 'ship_city',
                          state     => 'ship_state',
                          zip       => 'ship_zip',
                          country   => 'ship_country',
                        );

  tie my %customer, 'Tie::IxHash',
    $self->revmap_fields(
                          type           => 'customer_org',
                          id             => 'customer_id',
                          email          => 'email',
                          phoneNumber    => 'phone',
                          faxNumber      => 'fax',
                          driversLicense => \%drivers,
                          taxid          => 'customer_ssn',
                        );

  tie my %sub, 'Tie::IxHash',
    $self->revmap_fields(
                          name            => 'subscription_name',
                          paymentSchedule => \%schedule,
                          amount          => 'amount',
                          trialAmount     => 'trialamount',
                          payment         => \%payment,
                          order           => \%order,
                          customer        => \%customer,
                          billTo          => \%billto,
                          shipTo          => \%shipto,
                        );


  tie my %req, 'Tie::IxHash',
    $self->revmap_fields (
                           merchantAuthentication => \%merchant,
                           subscriptionId         => 'subscription',
                           subscription           => \%sub,
                         );

  my $ns = "AnetApi/xml/v1/schema/AnetApiSchema.xsd";
  my $post_data;
  my $writer = new XML::Writer( OUTPUT      => \$post_data,
                                DATA_MODE   => 1,
                                DATA_INDENT => 1,
                                ENCODING    => 'utf-8',
                              );
  $writer->xmlDecl();
  $writer->startTag($self->{_content}->{action}, 'xmlns', $ns);
  foreach ( keys ( %req ) ) {
    $self->_xmlwrite($writer, $_, $req{$_});
  }
  $writer->endTag($self->{_content}->{action});
  $writer->end();

  if ($self->test_transaction()) {
    $self->server('apitest.authorize.net');
  }

  warn $post_data if $DEBUG;
  my($page,$server_response,%headers) =
    $self->https_post( { 'Content-Type' => 'text/xml' }, $post_data);

  #trim leading (4?) characters of unknown origin not in spec
  $page =~ s/^(.*?)</</;
  my $garbage=$1;
  warn "Trimmed $garbage from response page.\n" if $DEBUG;

  warn $page if $DEBUG;

  my $response;
  my $message;
  if ($server_response =~ /200/){
    $response = XMLin($page);
    if (ref($response->{messages}->{message}) eq 'ARRAY') {
      $message = $response->{messages}->{message}->[0];
    }else{
      $message = $response->{messages}->{message};
    }
  }else{
    $response->{messages}->{resultCode} = "Server Failed";
    $message->{code} = $server_response;
  }

  $self->server_response($page);
  $self->order_number($response->{subscriptionId});
  $self->result_code($message->{code});
  $self->error_message($message->{text});

  if($response->{messages}->{resultCode} eq "Ok" ) {
      $self->is_success(1);
  } else {
      $self->is_success(0);
      unless ( $self->error_message() ) { #additional logging information
        $self->error_message(
          "(HTTPS response: $server_response) ".
          "(HTTPS headers: ".
            join(", ", map { "$_ => ". $headers{$_} } keys %headers ). ") ".
          "(Raw HTTPS content: $page)"
        );
      }
  }
}

1;
__END__

=head1 NAME

Business::OnlinePayment::AuthorizeNet::ARB - AuthorizeNet ARB backend for Business::OnlinePayment

=head1 AUTHOR

Jeff Finucane, authorizenetarb@weasellips.com

=head1 SEE ALSO

perl(1). L<Business::OnlinePayment>.

=cut



AuthorizeNet.pm
package Business::OnlinePayment::AuthorizeNet;

use strict;
use Carp;
use Business::OnlinePayment;
use vars qw($VERSION @ISA $me);

@ISA = qw(Business::OnlinePayment);
$VERSION = '3.22';
$me = 'Business::OnlinePayment::AuthorizeNet';

sub set_defaults {
    my $self = shift;

    $self->build_subs(qw( order_number md5 avs_code cvv2_response
                          cavv_response
                     ));
}

sub _map_processor {
    my($self) = @_;

    my %content = $self->content();
    my %processors = ('recurring authorization'          => 'ARB',
                      'modify recurring authorization'   => 'ARB',
                      'cancel recurring authorization'   => 'ARB',
                     );
    $processors{lc($content{'action'})} || 'AIM';
}

sub submit {
    my($self) = @_;

    my $processor = $me. "::". $self->_map_processor();

    eval "use $processor";
    croak("unknown processor $processor ($@)") if $@;
    
    my $object = bless $self, $processor;
    $object->set_defaults();
    $object->submit();
    bless $self, $me;
}

1;
__END__

=head1 NAME

Business::OnlinePayment::AuthorizeNet - AuthorizeNet backend for Business::OnlinePayment

=head1 SYNOPSIS

  use Business::OnlinePayment;

  ####
  # One step transaction, the simple case.
  ####

  my $tx = new Business::OnlinePayment("AuthorizeNet");
  $tx->content(
      type           => 'VISA',
      login          => 'testdrive',
      password       => '', #password or transaction key
      action         => 'Normal Authorization',
      description    => 'Business::OnlinePayment test',
      amount         => '49.95',
      invoice_number => '100100',
      customer_id    => 'jsk',
      first_name     => 'Jason',
      last_name      => 'Kohles',
      address        => '123 Anystreet',
      city           => 'Anywhere',
      state          => 'UT',
      zip            => '84058',
      card_number    => '4007000000027',
      expiration     => '09/02',
      cvv2           => '1234', #optional
      referer        => 'http://valid.referer.url/',
  );
  $tx->submit();

  if($tx->is_success()) {
      print "Card processed successfully: ".$tx->authorization."\n";
  } else {
      print "Card was rejected: ".$tx->error_message."\n";
  }

  ####
  # Two step transaction, authorization and capture.
  # If you don't need to review order before capture, you can
  # process in one step as above.
  ####

  my $tx = new Business::OnlinePayment("AuthorizeNet");
  $tx->content(
      type           => 'VISA',
      login          => 'testdrive',
      password       => '',  #password or transaction key
      action         => 'Authorization Only',
      description    => 'Business::OnlinePayment test',
      amount         => '49.95',
      invoice_number => '100100',
      customer_id    => 'jsk',
      first_name     => 'Jason',
      last_name      => 'Kohles',
      address        => '123 Anystreet',
      city           => 'Anywhere',
      state          => 'UT',
      zip            => '84058',
      card_number    => '4007000000027',
      expiration     => '09/02',
      cvv2           => '1234', #optional
      referer        => 'http://valid.referer.url/',
  );
  $tx->submit();

  if($tx->is_success()) {
      # get information about authorization
      $authorization = $tx->authorization
      $ordernum = $tx->order_number;
      $avs_code = $tx->avs_code; # AVS Response Code
      $cvv2_response = $tx->cvv2_response; # CVV2/CVC2/CID Response Code
      $cavv_response = $tx->cavv_response; # Cardholder Authentication
                                           # Verification Value (CAVV) Response
                                           # Code

      # now capture transaction
      my $capture = new Business::OnlinePayment("AuthorizeNet");

      $capture->content(
          type           => 'CC',
          action         => 'Post Authorization',
          login          => 'YOURLOGIN
          password       => 'YOURPASSWORD', #or transaction key
          order_number   => $ordernum,
          amount         => '49.95',
      );

      $capture->submit();

      if($capture->is_success()) { 
          print "Card captured successfully: ".$capture->authorization."\n";
      } else {
          print "Card was rejected: ".$capture->error_message."\n";
      }

  } else {
      print "Card was rejected: ".$tx->error_message."\n";
  }

  ####
  # One step subscription, the simple case.
  ####

  my $tx = new Business::OnlinePayment("AuthorizeNet::ARB");
  $tx->content(
      type           => 'CC',
      login          => 'testdrive',
      password       => 'testpass', #or transaction key
      action         => 'Recurring Authorization',
      interval       => '7 days',
      start          => '2008-3-10',
      periods        => '16',
      amount         => '99.95',
      trialperiods   => '4',
      trialamount    => '0',
      description    => 'Business::OnlinePayment test',
      invoice_number => '1153B33F',
      customer_id    => 'vip',
      first_name     => 'Tofu',
      last_name      => 'Beast',
      address        => '123 Anystreet',
      city           => 'Anywhere',
      state          => 'GA',
      zip            => '84058',
      card_number    => '4111111111111111',
      expiration     => '09/02',
  );
  $tx->submit();

  if($tx->is_success()) {
      print "Card processed successfully: ".$tx->order_number."\n";
  } else {
      print "Card was rejected: ".$tx->error_message."\n";
  }
  my $subscription = $tx->order_number


  ####
  # Subscription change.   Modestly more complicated.
  ####

  $tx->content(
      type           => 'CC',
      subscription   => '99W2C',
      login          => 'testdrive',
      password       => 'testpass', #or transaction key
      action         => 'Modify Recurring Authorization',
      interval       => '7 days',
      start          => '2008-3-10',
      periods        => '16',
      amount         => '29.95',
      trialperiods   => '4',
      trialamount    => '0',
      description    => 'Business::OnlinePayment test',
      invoice_number => '1153B340',
      customer_id    => 'vip',
      first_name     => 'Tofu',
      last_name      => 'Beast',
      address        => '123 Anystreet',
      city           => 'Anywhere',
      state          => 'GA',
      zip            => '84058',
      card_number    => '4111111111111111',
      expiration     => '09/02',
  );
  $tx->submit();

  if($tx->is_success()) {
      print "Update processed successfully."\n";
  } else {
      print "Update was rejected: ".$tx->error_message."\n";
  }
  $tx->content(
      subscription   => '99W2D',
      login          => 'testdrive',
      password       => 'testpass', # or transaction key
      action         => 'Cancel Recurring Authorization',
  );
  $tx->submit();

  ####
  # Subscription cancellation.   It happens.
  ####

  if($tx->is_success()) {
      print "Cancellation processed successfully."\n";
  } else {
      print "Cancellation was rejected: ".$tx->error_message."\n";
  }


=head1 SUPPORTED TRANSACTION TYPES

=head2 CC, Visa, MasterCard, American Express, Discover

Content required: type, login, password, action, amount, first_name, last_name, card_number, expiration.

=head2 Check

Content required: type, login, password, action, amount, first_name, last_name, account_number, routing_code, bank_name (non-subscription), account_type (subscription), check_type (subscription).

=head2 Subscriptions

Additional content required: interval, start, periods.

=head1 DESCRIPTION

For detailed information see L<Business::OnlinePayment>.

=head1 METHODS AND FUNCTIONS

See L<Business::OnlinePayment> for the complete list. The following methods either override the methods in L<Business::OnlinePayment> or provide additional functions.  

=head2 result_code

Returns the response reason code (from the message.code field for subscriptions).

=head2 error_message

Returns the response reason text (from the message.text field for subscriptions.

=head2 server_response

Returns the complete response from the server.

=head1 Handling of content(%content) data:

=head2 action

The following actions are valid

  normal authorization
  authorization only
  credit
  post authorization
  void
  recurring authorization
  modify recurring authorization
  cancel recurring authorization

=head2 interval

  Interval contains a number of digits, whitespace, and the units of days or months in either singular or plural form.
  

=head1 Setting AuthorizeNet ARB parameters from content(%content)

The following rules are applied to map data to AuthorizeNet ARB parameters
from content(%content):

      # ARB param => $content{<key>}
      merchantAuthentication
        name                     =>  'login',
        transactionKey           =>  'password',
      subscription
        paymentSchedule
          interval
            length               => \( the digits in 'interval' ),
            unit                 => \( days or months gleaned from 'interval' ),          startDate              => 'start',
          totalOccurrences       => 'periods',
          trialOccurrences       => 'trialperiods',
        amount                   => 'amount',
        trialAmount              => 'trialamount',
        payment
          creditCard
            cardNumber           => 'card_number',
            expiration           => \( $year.'-'.$month ), # YYYY-MM from 'expiration'
          bankAccount
            accountType          => 'account_type',
            routingNumber        => 'routing_code',
            accountNumber        => 'account_number,
            nameOnAccount        => 'name',
            bankName             => 'bank_name',
            echeckType           => 'check_type',
        order
          invoiceNumber          => 'invoice_number',
          description            => 'description',
        customer
          type                   => 'customer_org',
          id                     => 'customer_id',
          email                  => 'email',
          phoneNumber            => 'phone',
          faxNumber              => 'fax',
          driversLicense
            number               => 'license_num',
            state                => 'license_state',
            dateOfBirth          => 'license_dob',
          taxid                  => 'customer_ssn',
        billTo
          firstName              => 'first_name',
          lastName               => 'last_name',
          company                => 'company',
          address                => 'address',
          city                   => 'city',
          state                  => 'state',
          zip                    => 'zip',
          country                => 'country',
        shipTo
          firstName              => 'ship_first_name',
          lastName               => 'ship_last_name',
          company                => 'ship_company',
          address                => 'ship_address',
          city                   => 'ship_city',
          state                  => 'ship_state',
          zip                    => 'ship_zip',
          country                => 'ship_country',

=head1 NOTES

Use your transaction key in the password field.

Unlike Business::OnlinePayment or pre-3.0 verisons of
Business::OnlinePayment::AuthorizeNet, 3.1 requires separate first_name and
last_name fields.

Business::OnlinePayment::AuthorizeNet uses Authorize.Net's "Advanced
Integration Method (AIM) (formerly known as ADC direct response)" and
"Automatic Recurring Billing (ARB)", sending a username and password (or
transaction key as password) with every transaction.  Therefore,
Authorize.Net's referrer "security" is not necessary.  In your Authorize.Net
interface at https://secure.authorize.net/ make sure the list of allowable
referers is blank.  Alternatively, set the B<referer> field in the transaction
content.

To settle an authorization-only transaction (where you set action to
'Authorization Only'), submit the nine-digit transaction id code in
the field "order_number" with the action set to "Post Authorization".
You can get the transaction id from the authorization by calling the
order_number method on the object returned from the authorization.
You must also submit the amount field with a value less than or equal
to the amount specified in the original authorization.

For the subscription actions an authorization code is never returned by
the module.  Instead it returns the value of subscriptionId in order_number.
This is the value to use for changing or cancelling subscriptions.

Authorize.Net has turned address verification on by default for all merchants
since 2002.  If you do not have valid address information for your customer
(such as in an IVR application), you must disable address verification in the
Merchant Menu page at https://secure.authorize.net/ so that the transactions
aren't denied due to a lack of address information.

=head1 COMPATIBILITY

This module implements Authorize.Net's API using the Advanced Integration
Method (AIM) version 3.1, formerly known as ADC Direct Response and the 
Automatic Recurring Billing version 1.0 using the XML interface.  See
http://www.authorize.net/support/AIM_guide.pdf and http://www.authorize.net/support/ARB_guide.pdf for details.

=head1 AUTHORS

Original author: Jason Kohles, jason@mediabang.com

Ivan Kohler <ivan-authorizenet@freeside.biz> updated it for Authorize.Net
protocol 3.0/3.1 and is the current maintainer.  Please see the next section
for for information on contributing.

Jason Spence <jspence@lightconsulting.com> contributed support for separate
Authorization Only and Post Authorization steps and wrote some docs.
OST <services@ostel.com> paid for it.

Jeff Finucane <authorizenetarb@weasellips.com> added the ARB support.
ARB support sponsored by Plus Three, LP. L<http://www.plusthree.com>.

T.J. Mather <tjmather@maxmind.com> sent a number of CVV2 patches.

Mike Barry <mbarry@cos.com> sent in a patch for the referer field and a fix for
ship_company.

Yuri V. Mkrtumyan <yuramk@novosoft.ru> sent in a patch to add the void action.

Paul Zimmer <AuthorizeNetpm@pzimmer.box.bepress.com> sent in a patch for
card-less post authorizations.

Daemmon Hughes <daemmon@daemmonhughes.com> sent in a patch for "transaction
key" authentication as well support for the recurring_billing flag and the md5
method that returns the MD5 hash which is returned by the gateway.

Steve Simitzis contributed a patch for better compatibility with
eProcessingNetwork's AuthorizeNet compatability mode.

Michael G. Schwern contributed cleanups, test fixes, and more.

Erik Hollensbe implemented card-present data (track1/track2), the
duplicate_window parameter, and test fixes.

Paul Timmins added the check_number field.

Nate Nuss implemented the ("Additional Shipping Information (Level 2 Data)"
fields: tax, freight, duty, tax_exempt, po_number.

Michael Peters fixed a bug in email address handling.

Thomas Sibley <trs@bestpractical.com> wrote B:OP:AuthorizeNet::AIM::ErrorCodes
which was borged and used to provide more descriptive error messages.

=head1 CONTRIBUTIONS AND REPOSITORY

Please send patches as unified diffs (diff -u) to (in order of preference):

=over 4

=item CPAN RT

http://rt.cpan.org/Public/Bug/Report.html?Queue=Business-OnlinePayment-AuthorizeNet

=item The bop-devel mailing list

http://420.am/cgi-bin/mailman/listinfo/bop-devel

=item Ivan

Ivan Kohler <ivan-authorizenet@freeside.biz>

=back

The code is available from our public CVS repository:

  export CVSROOT=":pserver:anonymous@cvs.freeside.biz:/home/cvs/cvsroot"
  cvs login
  # The password for the user `anonymous' is `anonymous'.
  cvs checkout Business-OnlinePayment-AuthorizeNet

Or on the web:

  http://freeside.biz/cgi-bin/viewvc.cgi/Business-OnlinePayment-AuthorizeNet/

=head1 A WORD FROM OUR SPONSOR

This module and the Business::OnlinePayment framework are maintained by by
Freeside Internet Services.  If you need a complete, open-source web-based
application to manage your customers, billing and trouble ticketing, please
visit http://freeside.biz/

=head1 COPYRIGHT & LICENSE

Copyright 2010 Freeside Internet Services, Inc.
Copyright 2008 Thomas Sibley
All rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 SEE ALSO

perl(1). L<Business::OnlinePayment>.

=cut

